### PR TITLE
Propagate locale when using app initiated registration URL

### LINKED ISSUE #13505: locale attribute not set after registration
### Describe the bug

When a user creates a new account using the direct registration link of our application the user's locale is not saved as an attribute. If the user explicitly switches the current language by the language dropdown, the attribute is set, even if they just switch to english and back to german.

See also [this SO question](https://stackoverflow.com/questions/66208538/keycloak-locale-not-set-after-registration) describing the very same behaviour.

The registration link in our app is like this

https://<server>/auth/realms/<realm>/protocol/openid-connect/registrations?client_id=<id>&response_type=code&ui_locales=de&kc_locale=de&redirect_uri=<callback_url>

as mentioned [here ](https://stackoverflow.com/a/53264417/2217962) and in the referred keycloak mailing list archive.

### Version

17.0.1

### Expected behavior

Every user should have a "locale" attribute after registration, containing the language used during registration.

### Actual behavior

"locale" attribute is not set at all.

### How to Reproduce?

* Start keycloak
* Use registration link from above
* Register a new user
* Go to user detail page - locale dropdown is empty
* Read user attrbutes - locale is not set

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by thomasdarimont**
Fixes #13505

### CODE REVIEW

**Review by jonkoops — COMMENTED** [review]
Could you elaborate exactly how this fixed the issue? From what I can see the methods being called on the instance don't seem to use any state aside from what is passed into them, so I am not quite sure how using the static members statically would change the behaviour.

Also, do we have some tests for this code, if so could we add a test to verify this fix works as expected?

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java [hunk 1]
```diff
@@ -344,19 +346,20 @@ private Response buildAuthorizationCodeAuthorizationResponse() {
...
     private Response buildRegister() {
-        authManager.expireIdentityCookie(session);
+        AuthenticationManager.expireIdentityCookie(session);
 
         AuthenticationFlowModel flow = realm.getRegistrationFlow();
         String flowId = flow.getId();
 
         AuthenticationProcessor processor = createProcessor(authenticationSession, flowId, LoginActionsService.REGISTRATION_PATH);
         authenticationSession.setClientNote(APP_INITIATED_FLOW, LoginActionsService.REGISTRATION_PATH);
+        LocaleUtil.processLocaleParam(session, realm, authenticationSession);
```
Comment (thomasdarimont) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java:356]: This is the actual fix as it propagates the value from the KC_LOCALE_PARAM param to the proper auth session note.

Comment (thomasdarimont) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java:356]: The other two changes were just minor cleanups to get rid of warnings.

Comment (ahus1) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java:356]: @thomasdarimont - can you please amend an existing test or write a new test for this, so we don't break it in a future change? Thanks!

Comment (thomasdarimont) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java:355]: I minimized the change and added a small test.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedRegistrationTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Before
+    public void before() {
+        ApiUtil.removeUserByUsername(testRealm(), "test-user@localhost");
+    }
+
+    @Test
+    public void ensureLocaleParameterIsPropagatedDuringAppInitiatedRegistration() throws IOException, MessagingException {
+
+        String appInitiatedRegisterUrl = oauth.getLoginFormUrl();
```
Comment (jonkoops) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedRegistrationTest.java:37]: Perhaps we can add a new method for getting the registration URL?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedRegistrationTest.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        appInitiatedRegisterUrl = appInitiatedRegisterUrl.replace("openid-connect/auth", "openid-connect/registrations");
+
+        // add the kc_locale parameter
+        appInitiatedRegisterUrl += "&kc_locale=en";
```
Comment (jonkoops) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedRegistrationTest.java:41]: I'd prefer we add query params with some kind of utility class like `UriBuilder`, as it is less error prone.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedRegistrationTest.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        driver.navigate().to(appInitiatedRegisterUrl);
+
+        registerPage.assertCurrent();
+        registerPage.register("first", "last", "test-user@localhost", "test-user", "test","test");
+
+        appPage.assertCurrent();
+
+        UserRepresentation user = testRealm().users().searchByEmail("test-user@localhost", true).get(0);
+        // ensure that the locale was set on the user
+        Assert.assertEquals("en", user.getAttributes().get("locale").get(0));
```
Comment (jonkoops) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedRegistrationTest.java:52]: Is English not the default locale assigned to a user? Perhaps using a more foreign locale would be a good way to test.

**Review by jonkoops — DISMISSED** [review]
Got some comments, but not enough to not approve this as-is.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedRegistrationTest.java (hunk 4 above)
Comment (GioviQ) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedRegistrationTest.java:52]: the default should be the current language selected in login form,

Comment (thomasdarimont) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedRegistrationTest.java:52]: The local user attribute does not exist if the locale is not propagated properly.

I initially tested this with a german locale "de" however, then `appPage.assertCurrent();` breaks as it tries to match the english translation with a german title.

Hence I kept using "en".

Comment (thomasdarimont) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedRegistrationTest.java:49]: I applied the other suggestions to the PR.

**Review by ahus1 — APPROVED** [review]

**Pull request merged**